### ISSUE TITLE
Recover partitioning after missing execution of run_maintainance

### ISSUE THREAD

**Issue by Blindstealer** [keithf4/pg_partman#56]
Hi, i'm using the 1.8.4 release of pg_partman.
I wanted to time partion a table on my postgresql 9.3 db with an interval of partitioning of one hour, and so i set up a cronjob to call run_maintenance two times every hour. Premake is set to 2. I set up keep table and keep index on part_config at false (i want to delete older child partition) . I'm testing this configuration on my pc.
My problem is that last night, before i turned off my pc last night, i had the following tables in my db:

mytable
mytable_p2015_04_26_1300
...
mytable_p2015_04_26_1900t
mytable_p2015_04_15_2000

This morning (2014-04-27 at 10:00) i launched run_maintenance and the older child tables have been destroyed (which is something expected) but newer tables have not been created, and the only table in my database is "mytable" (the parent table), and insert operation are performed directly on it.
I suppose that this is due to the fact that run_maintenance has not been called twice the partitioning interval (because my computer was turned off during this interval).
In the same database I have another table that is partitioned daily but run_maintenance correctly created the new child table and deleted the older.
There is any way to recover from this state?

Thank you for your help

**Comment by keithf4** [comment]
Will have to test out this scenario and get back to you

**Comment by keithf4** [comment]
What did you have the `retention` column itself set to for the hourly partition set.
My current thoughts right now are that the retention system somehow deleted all the child tables in the hourly partition set because they were all eligible to be dropped. But, the way `run_maintenance()` works, it should have caught up and created all the new partitions it needed first before dropping the old ones. The "run twice within the partition interval" recommendation is only there as a safeguard to make sure the partitions are created ahead of time properly. If you miss it, it should automatically catch up and create all the partitions that are needed.

I've got the test scenario below set up right now. I'll let this sit until tomorrow then run `run_maintenance()` on it and see what happens. I've been testing things with a quarter-hour partition setup for the last few hours though and things have been fine.
```
testing_hourly=# \d+ partman_test.time_taptest_table
                              Table "partman_test.time_taptest_table"
 Column |           Type           |       Modifiers        | Storage  | Stats target | Description 
--------+--------------------------+------------------------+----------+--------------+-------------
 col1   | integer                  | not null               | plain    |              | 
 col2   | text                     |                        | extended |              | 
 col3   | timestamp with time zone | not null default now() | plain    |              | 
Indexes:
    "time_taptest_table_pkey" PRIMARY KEY, btree (col1)
Triggers:
    time_taptest_table_part_trig BEFORE INSERT ON partman_test.time_taptest_table FOR EACH ROW EXECUTE PROCEDURE partman_test.time_taptest_table_part_trig_func()
Child tables: partman_test.time_taptest_table_p2015_04_29_1100,
              partman_test.time_taptest_table_p2015_04_29_1200,
              partman_test.time_taptest_table_p2015_04_29_1300,
              partman_test.time_taptest_table_p2015_04_29_1400,
              partman_test.time_taptest_table_p2015_04_29_1500,
              partman_test.time_taptest_table_p2015_04_29_1600,
              partman_test.time_taptest_table_p2015_04_29_1700,
              partman_test.time_taptest_table_p2015_04_29_1800,
              partman_test.time_taptest_table_p2015_04_29_1900

testing_hourly=# select * from partman.part_config;
-[ RECORD 1 ]--------+--------------------------------
parent_table         | partman_test.time_taptest_table
control              | col3
type                 | time-static
part_interval        | 01:00:00
constraint_cols      | «NULL»
premake              | 2
inherit_fk           | t
retention            | 5 hours
retention_schema     | «NULL»
retention_keep_table | f
retention_keep_index | f
datetime_string      | YYYY_MM_DD_HH24MI
use_run_maintenance  | t
jobmon               | t
undo_in_progress     | f
```

**Comment by keithf4** [comment]
To recover from this, if there are no child tables left in the partition set at all, just delete that partition table's entry in the `part_config` table and run `create_parent()` on it again.

**Comment by keithf4** [comment]
Looks like you're right. Not sure why it's doing that out of order. 
```
testing_hourly=# \dt partman_test.
                         List of relations
    Schema    |                Name                 | Type  | Owner 
--------------+-------------------------------------+-------+-------
 partman_test | time_taptest_table                  | table | keith
 partman_test | time_taptest_table_p2015_04_29_1100 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1200 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1300 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1400 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1500 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1600 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1700 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1800 | table | keith
 partman_test | time_taptest_table_p2015_04_29_1900 | table | keith
(10 rows)

testing_hourly=# select now();
              now              
-------------------------------
 2015-05-01 17:07:12.539099-04
(1 row)

Time: 0.886 ms
testing_hourly=# select partman.run_maintenance();
 run_maintenance 
-----------------
 
(1 row)

Time: 171.396 ms
testing_hourly=# \dt partman_test.
                 List of relations
    Schema    |        Name        | Type  | Owner 
--------------+--------------------+-------+-------
 partman_test | time_taptest_table | table | keith
(1 row)
```
I'll try and look into it. Ideally it should always leave at least the latest child table.

To reset things needs a bit more work also. You in addition to removing the config data, you have to remove the trigger & trigger function as well. The other alternative is to just manually make the latest child table and the maintenance process should pick up from there.
```
testing_hourly=# select * from partman.part_config;
-[ RECORD 1 ]--------+--------------------------------
parent_table         | partman_test.time_taptest_table
control              | col3
type                 | time-static
part_interval        | 01:00:00
constraint_cols      | «NULL»
premake              | 2
inherit_fk           | t
retention            | 5 hours
retention_schema     | «NULL»
retention_keep_table | f
retention_keep_index | f
datetime_string      | YYYY_MM_DD_HH24MI
use_run_maintenance  | t
jobmon               | t
undo_in_progress     | f

Time: 1.323 ms
testing_hourly=# delete from partman.part_config where parent_table = 'partman_test.time_taptest_table';
DELETE 1
Time: 9.389 ms

testing_hourly=# \d partman_test.time_taptest_table
          Table "partman_test.time_taptest_table"
 Column |           Type           |       Modifiers        
--------+--------------------------+------------------------
 col1   | integer                  | not null
 col2   | text                     | 
 col3   | timestamp with time zone | not null default now()
Indexes:
    "time_taptest_table_pkey" PRIMARY KEY, btree (col1)
Triggers:
    time_taptest_table_part_trig BEFORE INSERT ON partman_test.time_taptest_table FOR EACH ROW EXECUTE PROCEDURE partman_test.time_taptest_table_part_trig_func()

testing_hourly=# drop trigger time_taptest_table_part_trig on partman_test.time_taptest_table;
DROP TRIGGER
Time: 10.453 ms

testing_hourly=# drop function partman_test.time_taptest_table_part_trig_func();
DROP FUNCTION
Time: 8.799 ms

testing_hourly=# SELECT partman.create_parent('partman_test.time_taptest_table', 'col3', 'time-static', 'hourly');
 create_parent 
---------------
 t
(1 row)

Time: 318.793 ms

testing_hourly=# \d+ partman_test.time_taptest_table
                              Table "partman_test.time_taptest_table"
 Column |           Type           |       Modifiers        | Storage  | Stats target | Description 
--------+--------------------------+------------------------+----------+--------------+-------------
 col1   | integer                  | not null               | plain    |              | 
 col2   | text                     |                        | extended |              | 
 col3   | timestamp with time zone | not null default now() | plain    |              | 
Indexes:
    "time_taptest_table_pkey" PRIMARY KEY, btree (col1)
Triggers:
    time_taptest_table_part_trig BEFORE INSERT ON partman_test.time_taptest_table FOR EACH ROW EXECUTE PROCEDURE partman_test.time_taptest_table_part_trig_func()
Child tables: partman_test.time_taptest_table_p2015_05_01_1300,
              partman_test.time_taptest_table_p2015_05_01_1400,
              partman_test.time_taptest_table_p2015_05_01_1500,
              partman_test.time_taptest_table_p2015_05_01_1600,
              partman_test.time_taptest_table_p2015_05_01_1700,
              partman_test.time_taptest_table_p2015_05_01_1800,
              partman_test.time_taptest_table_p2015_05_01_1900,
              partman_test.time_taptest_table_p2015_05_01_2000,
              partman_test.time_taptest_table_p2015_05_01_2100
```

**Comment by keithf4** [comment]
Believe I've gotten this fixed in next version now. Issue was actually with catching up in general when maintenance hadn't been run for a while, not just with the retention system. Created a partition set last friday and waited until today to see if it caught up.

Will release next version as soon as I get some more testing done.
```
testing_hourly=# \dt partman_test.
                         List of relations
    Schema    |                Name                 | Type  | Owner 
--------------+-------------------------------------+-------+-------
 partman_test | time_taptest_table                  | table | keith
 partman_test | time_taptest_table_p2015_05_08_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1700 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1800 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1900 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2000 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2100 | table | keith
(10 rows)

testing_hourly=# select now();
              now              
-------------------------------
 2015-05-11 12:39:52.234828-04
(1 row)

Time: 1.540 ms

testing_hourly=# select partman.run_maintenance('partman_test.time_taptest_table');
 run_maintenance 
-----------------
 
(1 row)

Time: 1338.910 ms
testing_hourly=# \dt partman_test.
                         List of relations
    Schema    |                Name                 | Type  | Owner 
--------------+-------------------------------------+-------+-------
 partman_test | time_taptest_table                  | table | keith
 partman_test | time_taptest_table_p2015_05_08_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1700 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1800 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1900 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2000 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2100 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2200 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2300 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0000 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0100 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0200 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0300 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0400 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0500 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0600 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0700 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0800 | table | keith
 partman_test | time_taptest_table_p2015_05_09_0900 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1000 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1100 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1200 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1700 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1800 | table | keith
 partman_test | time_taptest_table_p2015_05_09_1900 | table | keith
 partman_test | time_taptest_table_p2015_05_09_2000 | table | keith
 partman_test | time_taptest_table_p2015_05_09_2100 | table | keith
 partman_test | time_taptest_table_p2015_05_09_2200 | table | keith
 partman_test | time_taptest_table_p2015_05_09_2300 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0000 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0100 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0200 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0300 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0400 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0500 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0600 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0700 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0800 | table | keith
 partman_test | time_taptest_table_p2015_05_10_0900 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1000 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1100 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1200 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1700 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1800 | table | keith
 partman_test | time_taptest_table_p2015_05_10_1900 | table | keith
 partman_test | time_taptest_table_p2015_05_10_2000 | table | keith
 partman_test | time_taptest_table_p2015_05_10_2100 | table | keith
 partman_test | time_taptest_table_p2015_05_10_2200 | table | keith
 partman_test | time_taptest_table_p2015_05_10_2300 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0000 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0100 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0200 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0300 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0400 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0500 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0600 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0700 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0800 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0900 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1000 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1100 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1200 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1600 | table | keith
(77 rows)
```

**Comment by keithf4** [comment]
Just to be sure, tested with retention in place as well and all seems well
```
testing_hourly=# \dt partman_test.
                         List of relations
    Schema    |                Name                 | Type  | Owner 
--------------+-------------------------------------+-------+-------
 partman_test | time_taptest_table                  | table | keith
 partman_test | time_taptest_table_p2015_05_08_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1700 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1800 | table | keith
 partman_test | time_taptest_table_p2015_05_08_1900 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2000 | table | keith
 partman_test | time_taptest_table_p2015_05_08_2100 | table | keith
(10 rows)

testing_hourly=# select * from partman.part_config;
-[ RECORD 1 ]--------+--------------------------------
parent_table         | partman_test.time_taptest_table
control              | col3
partition_type       | time
partition_interval   | 01:00:00
constraint_cols      | «NULL»
premake              | 4
inherit_fk           | t
retention            | 05:00:00
retention_schema     | «NULL»
retention_keep_table | f
retention_keep_index | t
datetime_string      | YYYY_MM_DD_HH24MI
use_run_maintenance  | t
jobmon               | t
undo_in_progress     | f

Time: 1.584 ms
testing_hourly=# select now();
              now              
-------------------------------
 2015-05-11 13:26:02.289978-04
(1 row)

Time: 1.464 ms
testing_hourly=# select partman.run_maintenance();
 run_maintenance 
-----------------
 
(1 row)

Time: 1274.200 ms
testing_hourly=# \dt partman_test.
                         List of relations
    Schema    |                Name                 | Type  | Owner 
--------------+-------------------------------------+-------+-------
 partman_test | time_taptest_table                  | table | keith
 partman_test | time_taptest_table_p2015_05_11_0800 | table | keith
 partman_test | time_taptest_table_p2015_05_11_0900 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1000 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1100 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1200 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1300 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1400 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1500 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1600 | table | keith
 partman_test | time_taptest_table_p2015_05_11_1700 | table | keith
(11 rows)

```

**Comment by keithf4** [comment]
This should be fixed with v1.8.5. Let me know if you're still having any issues.

**Comment by Blindstealer** [comment]
Thank you for the help! 
Tomorrow I'll update partman and let you know if now works!
